File: hw3/classifier/perceptron.m
function [w,theta,error]=perceptron(x,y,eta,cycles,R)
    [m,n] = size(x);
    
    if isempty(R)
        w = zeros(1,n);
        theta = 0;
        error = zeros(1,m*cycles);
        for j = 1:cycles
            for i = 1:m
                if (dot(w,x(i,:))+theta) * y(i) <= 0
                    w = w + eta * y(i) * x(i,:);
                    theta = theta + eta * y(i);
                    error(1,(j-1)*m + i) = 1;
                end
            end
        end
    else
        w = zeros(1,n);
        theta = 0;
        error = 0;
        correct = 0; 
        while correct < R
            for i = 1:m
                if (dot(w,x(i,:))+theta) * y(i) <= 0
                    error = error + 1;
                    correct = 0;
                    
                    w = w + eta * y(i) * x(i,:);
                    theta = theta + eta * y(i);
                else
                    correct = correct + 1;
                    if (correct >= R)
                        return;
                    end
                end
            end
        end
    end
end

File: hw4/perceptron.py
from __future__ import print_function

import sys
import numpy as np

from abc import ABCMeta, abstractmethod
import data_loader


ETA = "eta"
GAMMA = "gamma"
ALPHA = "alpha"


class BaseLearner(object):
    __metaclass__ = ABCMeta

    def __init__(self, dimensions, parameters):
        self.parameters = parameters
        self.weight_vector = np.zeros(dimensions)
        self.theta = 0

    def predictionScore(self, x, y):
        y = 2*y - 1 #This converts {0,1} to {-1,1}
        return y * (np.dot(self.weight_vector, x) + self.theta)

    @abstractmethod
    def runIterationOnOneExample(self, x, y): pass

    def evaluate(self, data):
        num_instances = len(data)

        mistakes = 0.0
        for i in range(num_instances):
            if self.predictionScore(data[i][0], data[i][1]) <= 0:
                mistakes += 1

        return (num_instances-mistakes)/num_instances

    def train(self, data, num_rounds=100):
        num_instances = len(data)

        for iteration in range(num_rounds):
            update_count = 0
            for i in range(num_instances):
                update_count += self.runIterationOnOneExample(data[i][0], data[i][1])
            
        return self.evaluate(data)

    def train_with_learning_curve(self, data, num_rounds=100):
        num_instances = len(data)
        learning_curve_data = []
        for iteration in range(num_rounds):
            update_count = 0
            for i in range(num_instances):
                update_count += self.runIterationOnOneExample(data[i][0], data[i][1])
            learning_curve_data.append((iteration+1, self.evaluate(data)))
            
        return learning_curve_data




class Perceptron(BaseLearner):
    """
    Perceptron implementation:

    Usage:

    learner = Perceptron(weight_vector_dimension)
    """


    def __init__(self, dimensions):
        BaseLearner.__init__(self, dimensions, {})

    def runIterationOnOneExample(self, x, y):
        y = 2*y - 1 #This converts {0,1} to {-1, 1}
        if self.predictionScore(x, y) <= 0:
            self.weight_vector = self.weight_vector + (y * x)
            self.theta = self.theta + y
            return 1

        return 0

